helper/resource/testing_new.go

```diff
@@ -253,34 +254,107 @@ func runNewTest(ctx context.Context, t testing.T, c TestCase, helper *plugintest
 		if step.RefreshState {
 			logging.HelperResourceTrace(ctx, "TestStep is RefreshState mode")
 
-			err := testStepNewRefreshState(ctx, t, wd, step, providers)
+			testStepNewRefreshStateResponse, err := testStepNewRefreshState(ctx, t, wd, step, providers)
+			refreshTfJSONDiags := testStepNewRefreshStateResponse.tfJSONDiags
+
 			if step.ExpectError != nil {
 				logging.HelperResourceDebug(ctx, "Checking TestStep ExpectError")
+
 				if err == nil {
 					logging.HelperResourceError(ctx,
 						"Error running refresh: expected an error but got none",
 					)
 					t.Fatalf("Step %d/%d error running refresh: expected an error but got none", stepNumber, len(c.Steps))
 				}
-				if !step.ExpectError.MatchString(err.Error()) {
+
+				errorFound := step.ExpectError.MatchString(err.Error())
+				jsonErrorFound := refreshTfJSONDiags.Contains(step.ExpectError, tfjson.DiagnosticSeverityError)
+
+				errorOutput := []string{err.Error()}
+
+				for _, v := range refreshTfJSONDiags.Errors() {
+					b, err := json.Marshal(v)
+					if err != nil {
+						t.Errorf("could not marshal tfjson diagnostic: %s", err)
+					}
+
+					errorOutput = append(errorOutput, string(b))
+				}
+
+				if !errorFound && !jsonErrorFound {
 					logging.HelperResourceError(ctx,
 						fmt.Sprintf("Error running refresh: expected an error with pattern (%s)", step.ExpectError.String()),
-						map[string]interface{}{logging.KeyError: err},
+						map[string]interface{}{logging.KeyError: strings.Join(errorOutput, "")},
 					)
-					t.Fatalf("Step %d/%d error running refresh, expected an error with pattern (%s), no match on: %s", stepNumber, len(c.Steps), step.ExpectError.String(), err)
+					t.Fatalf("Step %d/%d error running refresh, expected an error with pattern (%s), no match on: %s", stepNumber, len(c.Steps), step.ExpectError.String(), strings.Join(errorOutput, "\n"))
 				}
 			} else {
 				if err != nil && c.ErrorCheck != nil {
 					logging.HelperResourceDebug(ctx, "Calling TestCase ErrorCheck")
+
+					// This is a "best effort" to supply ErrorCheck with the contents of the
+					// error diagnostics returned from running the Terraform command when
+					// executed with the -json flag. If ErrorCheck functions are using
+					// regexp to perform matches that include line breaks then these will
+					// likely no longer behave as expected and therefore potentially
+					// represent a breaking change.
```

Comment:
For others benefit, the `TestCase.ErrorCheck` functionality is the main wrinkle that prevents us from immediately being able to use Terraform's machine readable output (JSON view/UI).

`ErrorCheck` functionality, currently used in many popular Terraform provider acceptance tests, was intended to serve these use cases:

- Skipping tests based on known errors (e.g. an API error indicating that a feature is not supported against an endpoint variant)
- Rewriting errors for clarity
- Ignoring known errors

Today the testing code is reliant on terraform-exec's `error` returns to include the data needed for `TestCase.ErrorCheck` and `TestStep.ExpectError` functionality to work. Warning diagnostics are not passed back the same way, since they are not errors and should not stop program execution. In order to capture similar data for warnings, the testing code would either need to perform error-prone (and likely version dependent) scraping of the human-readable output or switch to the machine-readable output. The machine-readable output means that the `error` from terraform-exec needs to be fundamentally treated differently and trying to convert machine-readable error diagnostics to be functionally equivalent to the prior `error` would be a non-trivial exercise that's not guaranteed to fully keep backwards compatibility.

An approach to this situation may be to prevent defining both `TestCase.ErrorCheck` and the new `TestStep.ExpectWarning` in the same `TestCase`, and setting up the testing execution to continue using the human-readable output where `ErrorCheck` is used. This is a non-starter for providers such as the Terraform AWS Provider though, which defines `ErrorCheck` functionality on _all_ tests, meaning `ExpectWarning` could never be used.

Definitely open to ideas here, but trying to work around this further is risky at best.